plugins/slick.state.js

```diff
@@ -70,13 +70,20 @@
       save();
     }
 
-    function save() {
+    function save(userData) {
       if (_cid && _store) {
         var state = {
           sortcols: getSortColumns(),
           viewport: _grid.getViewport(),
-          columns: getColumns()
+          columns: getColumns(),
+          userData: null
         };
+
+        if (userData !== undefined && !(userData instanceof jQuery.Event))
```

Comment:
I'm not actually using this plugin in my own libs (I have my own Grid State Service in my libs which is a lot more complex) but the only thing I see here is what would happen if jQuery Event is no longer jQuery (I mean what if that portion is rewritten in plain JS Event, then I guess this code would no longer work)

Reply:
if the save event is legitimately provided with parameters ```(e, args)``` at times, then it should really have those added to its function declaration, and it would make sense for ```myData``` to be either a member of ```args```, or a third additional parameter.  However I'm not sure that ```(e, args)``` *should* really be provided to ```save()``` - if it's not then the parameters should be removed in the calling code and the problem disappears.

Reply:
> if the save event is legitimately provided with parameters `(e, args)` at times, then it should really have those added to its function declaration, and it would make sense for `myData` to be either a member of `args`, or a third additional parameter. However I'm not sure that `(e, args)` _should_ really be provided to `save()` - if it's not then the parameters should be removed in the calling code and the problem disappears.

as - hopefully - visible in the following image, sorting the columns causes a call to `save` function, as it changes the state of the grid. unfortunately, the way the subscribe/notify system is implemented, I see no easy way around not passing `e, args` to the function. moreover, it's not possible to pass the current `userData` to the trigger function, as the calling function is in `slick.grid.js`. so, `userData` cannot be set as a member of the `args` and it also doesn't make sense, because the change was sorting of the columns and its `args` are the related to that action and shall not contain `userData`.

```js
trigger(self.onSort, {
       multiColumnSort: true,
       sortCols: $.map(sortColumns, function(col) {
       var column = getColumnById(col.columnId, false);
      return {columnId: column && column.id, sortCol: column, sortAsc: col.sortAsc };
     })
  }, 
e);
```

![grid-menu-state](https://user-images.githubusercontent.com/2896170/131632784-0b1147b3-b9c0-4330-9e34-34286002d97c.gif)

Reply:
> it seems to be roughly the same technique to check for `instanceof Event` as shown in this [SO Answer](https://stackoverflow.com/a/27066820/1212166). So perhaps you change code it like this?
> 
> ```js
>  if (userData !== undefined && !(userData instanceof jQuery.Event || userData instanceof Event))
> ```

gladly I'll do so.

how about my proposed solution with the `setUserData` which stores `userData` for only the next `save` function call and then throws it away?

Reply:
better yet, 3 functions `setUserData`, `getUserData` and `revertUserData` would be the nicest way to implement it. `setUserData` sets the `userData` but doesn't store it in localStorage. `userData` is only saved to localStorage only using `save`. `getUserData` gets the value from the current `userData` (i.e. first choice is the value set by `setUserData`, if not set, then the value from state. `revertUserData` reverts currently set `userData` to the value saved in state...

Reply:
I don't have any objections since I'm not personally using this plugin, the get/set user data is fine by me, I'm not sure about the revert (did you mean reset?). In mine I only have getter functions (getCurrentColumns, getCurrentFilters, ...), I don't have any setters since it's supposed to return the grid state and I'm not providing any ways of providing extra data. You're a big user of this plugin so go ahead and change it like you think that it might be useful to others.

Reply:
I just saw your animated gif and just as a side note, I added the `onBeforeSort` event couple days ago which was released in the new version. I'm not sure if it really changes anything in your code but if the event returns `false` and is bubbling prevented, then it won't execute the `onSort`, you should still be ok to save only with the `onSort`
https://github.com/6pac/SlickGrid/blob/5fa4d8c3dfc8150f8e9525f5a5b2c1905b250bb5/slick.grid.js#L1404-L1426

Reply:
well, the new way of setting `userData` (using `setUserData` instead of `save` has eliminated conflicts here:

* as the signature of the save function will stay the same, all the calls to save the state will behave as before. 
* if someone wants to add `userData`, they will need to use `state.setUserData({...})`. they will only be saved IF the user calls `save` function directly or indirectly (using triggers or such). `setUserData` does not trigger the save function; allowing the user to manipulate `userData` before the actual `save`ing function call...